src/main/scala/chisel3/util/MemInterface.scala

```diff
@@ -0,0 +1,337 @@
+    macro MemTransform.memInterface_withMask[T]
+
+  /** Generates a [[SyncReadMem]] wrapper connected to an explicit number of read, write,
+    * and read/write ports, with masking capability on all write and read/write ports
+    *
+    * @param size The desired size of the inner `SyncReadMem`
+    * @tparam T The data type of the memory element
+    * @param numRd The number of desired read ports, >= 0
+    * @param numWr The number of desired write ports, >= 0
+    * @param numRdWr The number of desired read/write ports, >= 0
+    * @param clock The clock to bind to each generated port, which may be different from the implicit clock
+    *
+    * @return A new `MemInterface` wire containing the control signals for each instantiated port
+    * @note This does *not* return the wrapper module itself, you must interact with it using the returned bundle
+    */
+  def withMask[T <: Data](
+    size:    BigInt,
+    tpe:     T,
+    numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int,
+    clock:   Clock
+  )(
+    implicit evidence: T <:< Vec[_]
+  ): MemInterface[T] =
+    macro MemTransform.memInterface_withMaskClk[T]
+
+  /** @group SourceInfoTransformMacro */
+  def do_apply[T <: Data](
+    size:    BigInt,
+    tpe:     T,
+    numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int
+  )(
+    implicit sourceInfo: SourceInfo
+  ): MemInterface[T] = memInterface_impl(size, tpe)(numRd, numWr, numRdWr, Builder.forcedClock)
+
+  /** @group SourceInfoTransformMacro */
+  def do_apply[T <: Data](
+    size:    BigInt,
+    tpe:     T,
+    numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int,
+    clock:   Clock
+  )(
+    implicit sourceInfo: SourceInfo
+  ): MemInterface[T] = memInterface_impl(size, tpe)(numRd, numWr, numRdWr, clock)
+
+  /** @group SourceInfoTransformMacro */
+  def do_withMask[T <: Data](
+    size:    BigInt,
+    tpe:     T,
+    numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int
+  )(
+    implicit sourceInfo: SourceInfo,
+    evidence:            T <:< Vec[_]
+  ): MemInterface[T] = memInterface_withMask_impl(size, tpe)(numRd, numWr, numRdWr, Builder.forcedClock)
+
+  /** @group SourceInfoTransformMacro */
+  def do_withMask[T <: Data](
+    size:    BigInt,
+    tpe:     T,
+    numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int,
+    clock:   Clock
+  )(
+    implicit sourceInfo: SourceInfo,
+    evidence:            T <:< Vec[_]
+  ): MemInterface[T] = memInterface_withMask_impl(size, tpe)(numRd, numWr, numRdWr, clock)
+
+  /** @group SourceInfoTransformMacro */
+  private def memInterface_impl[T <: Data](
+    size:    BigInt,
+    tpe:     T
+  )(numRd:   Int,
+    numWr:   Int,
+    numRdWr: Int,
+    clock:   Clock
+  )(
+    implicit sourceInfo: SourceInfo
+  ): MemInterface[T] = {
+    val addrWidth = log2Up(size + 1)
+
+    val _wrappedMem = Module(new Module {
+      override def desiredName: String =
+        s"SyncReadMemWrapper_${MemInterface.portedness(numRd, numWr, numRdWr)}_${tpe.typeName}"
```

Comment:
Do we really want a level of module hierarchy here? If we do, this Module should at least be a named class, if we're trying to hide it from the user it can be private but I think it's better to avoid anonymous classes for this sort of thing.

I'm less sure if we want the level of hierarchy here. Part of the stated goal of an explicit memory API is that it may ultimately just directly instantiate FIRRTL mems rather than Chirrtl mems. Would we still want the layer of hierarchy in that case? Having this as an anonymous Module rather than a public class suggests to me we may be removing the level of hierarchy, but levels of module hierarchy are the canonical example of [Hyrum's Law](https://www.hyrumslaw.com) in Chisel--users shouldn't rely on the specific module hierarchy but they usually do--especially for memories. So I'm wondering if we should just not have this level of hierarchy here.